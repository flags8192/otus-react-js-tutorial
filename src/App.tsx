import React, { FC, useEffect, useState } from 'react';
import '@src/App.css';
import { GameField } from '@src/GameField';
import { CellInfo } from '@src/Cell/models';
import { getInverted } from '@src/Cell/utils';

const START_WIDTH = 3;
const START_HEIGHT = 3;

const App: FC = () => {
  const width = START_WIDTH;
  const height = START_HEIGHT;
  const showAll = false;

  const [data, setData] = useState<Array<CellInfo>>([]);

  useEffect(() => {
    const cellsNumber = width * height;
    const startCellState = showAll ? CellInfo.alive : CellInfo.dead;
    const newData: CellInfo[] = [];
    for (let i = 0; i < cellsNumber; i = i + 1) {
      newData.push(startCellState);
    }
    setData(newData);
  }, [width, height, showAll]);

  const handleClick = (cellId: number) => {
    if (cellId < 0 || cellId >= data.length) {
      return;
    }

    const newData = data.concat();
    newData[cellId] = getInverted(newData[cellId]);
    setData(newData);
  };

  return (
    <div className="App">
      <header className="App-header">
        <p>Game of Life</p>
      </header>
      <GameField data={data} onCellClick={handleClick} width={width} />
    </div>
  );
};

export default App;
