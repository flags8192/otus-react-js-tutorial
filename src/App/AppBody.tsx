import React, { FC } from 'react';
import styled from '@emotion/styled';
import GameField from '@src/modules/GameField';

const Root = styled.div`
  height: 80vh;
  position: relative;
`;

const AppBody: FC = () => {
  return (
    <Root role="gamePanel">
      <GameField />
    </Root>
  );
};

export default AppBody;
