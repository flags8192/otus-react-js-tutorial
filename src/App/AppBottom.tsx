import React, { FC, memo, useContext } from 'react';
import Panel from '@src/layout/Panel';
import styled from '@emotion/styled';
import store from '@src/store';

const Message = styled.p`
  margin: auto 20px;
`;

const AppBottom: FC = () => {
  const { state } = useContext(store);
  const { gameFieldPercentFilled } = state;

  return (
    <Panel role="bottomPanel">
      <Message role="messageResult">Результат:</Message>
      <Message role="messagePercentFilled">{`Процент заполнения: ${gameFieldPercentFilled}`}</Message>
    </Panel>
  );
};

export default memo(AppBottom);
