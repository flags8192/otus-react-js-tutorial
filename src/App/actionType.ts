const prefix = 'APP__';

export default {
  RESET: `${prefix}RESET`,
  SET_LEVEL: `${prefix}SET_LEVEL`,
};
