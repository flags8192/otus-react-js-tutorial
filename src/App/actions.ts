import actionType from './actionType';

const resetApp = () => ({
  type: actionType.RESET,
});

const setLevel = (gameLevel: string) => ({
  type: actionType.SET_LEVEL,
  payload: { gameLevel },
});

export { resetApp, setLevel };
