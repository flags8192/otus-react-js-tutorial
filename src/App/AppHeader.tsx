import React, { FC, memo, useContext } from 'react';
import Panel from '@src/layout/Panel';
import styled from '@emotion/styled';
import Button from '@src/components/Button';
import store from '@src/store';
import { resetApp, setLevel } from '@src/App/actions';
import Select from '@src/components/Select';

const Title = styled.p`
  margin: auto 20px;
`;

const AppHeader: FC = () => {
  const { dispatch } = useContext(store);

  const handleReset = () => {
    dispatch(resetApp());
  };

  const handleSelectLevel = (event: React.ChangeEvent<HTMLSelectElement>) => {
    const { value } = event.currentTarget;
    dispatch(setLevel(value));
  };

  return (
    <Panel role="topPanel">
      <Title>Lines</Title>
      <Button onClick={handleReset} title="Reset" />
      <Select onSelect={handleSelectLevel} />
    </Panel>
  );
};

export default memo(AppHeader);
