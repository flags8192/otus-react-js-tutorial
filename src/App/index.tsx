import React, { FC } from 'react';
import styled from '@emotion/styled';
import { Global, css } from '@emotion/core';

import AppHeader from '@src/App/AppHeader';
import AppBottom from '@src/App/AppBottom';
import AppBody from '@src/App/AppBody';

const Root = styled.div`
  overflow: hidden;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen', 'Ubuntu',
    'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue', sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
`;

const appGlobal = css`
  body {
    background: #282c34;
    color: white;
    margin: 0;
    padding: 0;
    min-height: '100vh';
    max-width: '100vw';
  }
`;

interface IUser {
  body: string;
  id: number;
  title: string;
  userId: number;
}

export type IUserList = Array<IUser>;

const App: FC = () => {
  return (
    <Root>
      <Global styles={appGlobal} />
      <AppHeader />
      <AppBody />
      <AppBottom />
    </Root>
  );
};

export default App;
