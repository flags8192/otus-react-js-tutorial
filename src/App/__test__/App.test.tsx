import React from 'react';
import { render, screen } from '@testing-library/react';

import App from '@src/App';
import Store from '@src/store';

test('render App', () => {
  const state = {
    gameLevel: '1',
    gameFieldSize: 3,
    gameFieldPercentFilled: 10,
    gameFieldData: [1, 0, 0, 0, 0, 0, 0, 0, 0],
    selectedCells: {},
  };
  const dispatch = jest.fn();
  const { asFragment } = render(
    <Store.Provider value={{ dispatch, state }}>
      <App />
    </Store.Provider>
  );
  expect(asFragment()).toMatchSnapshot();
  const topPanel = screen.getByRole(/topPanel/gi);
  expect(topPanel).toBeInTheDocument();
  const gamePanel = screen.getByRole(/gamePanel/gi);
  expect(gamePanel).toBeInTheDocument();
  const bottomPanel = screen.getByRole(/bottomPanel/gi);
  expect(bottomPanel).toBeInTheDocument();
});
