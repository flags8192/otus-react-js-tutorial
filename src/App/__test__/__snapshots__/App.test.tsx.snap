// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render App 1`] = `
<DocumentFragment>
  <div
    class="css-5simcc"
  >
    <div
      class="css-cogpxn"
      role="topPanel"
    >
      <p
        class="css-o9gtt2"
      >
        Lines
      </p>
      <button
        class="css-tv5zn1"
        role="button"
      >
        Reset
      </button>
      <select
        class="css-tv5zn1"
        name="select"
        role="select"
      >
        <option
          role="option"
          value="1"
        >
          Падаван
        </option>
        <option
          role="option"
          value="2"
        >
          Джедай
        </option>
        <option
          role="option"
          value="3"
        >
          Мастер-джедай
        </option>
        <option
          role="option"
          value="4"
        >
          Член Совета Силы
        </option>
      </select>
    </div>
    <div
      class="css-xt4orx"
      role="gamePanel"
    >
      <section
        class="css-hsjawn"
        role="cellsGrid"
      >
        <div
          class="css-es8rv7"
          role="cellContainer-0"
        >
          <div
            class="css-w6oewy"
            role="cellContent-0"
          />
        </div>
        <div
          class="css-2r18jt"
          role="cellContainer-1"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-2"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-3"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-4"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-5"
        />
        <div
          class="css-es8rv7"
          role="cellContainer-6"
        />
        <div
          class="css-2r18jt"
          role="cellContainer-7"
        />
        <div
          class="css-1jbr29x"
          role="cellContainer-8"
        />
        <div
          style="clear: both;"
        />
      </section>
    </div>
    <div
      class="css-cogpxn"
      role="bottomPanel"
    >
      <p
        class="css-o9gtt2"
        role="messageResult"
      >
        Результат:
      </p>
      <p
        class="css-o9gtt2"
        role="messagePercentFilled"
      >
        Процент заполнения: 10
      </p>
    </div>
  </div>
</DocumentFragment>
`;
